module/VuFind/src/VuFind/Db/Service/AuditEventService.php

```diff
@@ -0,0 +1,285 @@
+        if (null !== $user) {
+            $conditions[] = 'e.user = :user';
+            $params['user'] = $user;
+        }
+        if (null !== $username) {
+            $conditions[] = 'e.username = :username';
+            $params['username'] = $username;
+        }
+        if (null !== $clientIp) {
+            $conditions[] = 'e.clientIp = :clientIp';
+            $params['clientIp'] = $clientIp;
+        }
+        if (null !== $serverIp) {
+            $conditions[] = 'e.serverIp = :serverIp';
+            $params['serverIp'] = $serverIp;
+        }
+        if (null !== $serverName) {
+            $conditions[] = 'e.serverName = :serverName';
+            $params['serverName'] = $serverName;
+        }
+
+        if ($conditions) {
+            $dql .= ' WHERE ' . implode(' AND ', $conditions);
+        }
+
+        if ($sort) {
+            $sortFields = array_map(
+                fn ($s) => "e.$s",
+                $sort
+            );
+            $dql .= ' ORDER BY ' . implode(', ', $sortFields);
+        }
+
+        return $this->entityManager
+            ->createQuery($dql)
+            ->setParameters($params)
+            ->getResult();
+    }
+
+    /**
+     * Delete expired records. Allows setting a limit so that rows can be deleted in small batches.
+     *
+     * @param DateTime $dateLimit Date threshold of an "expired" record.
+     * @param ?int     $limit     Maximum number of rows to delete or null for no limit.
+     *
+     * @return int Number of rows deleted
+     */
+    public function deleteExpired(DateTime $dateLimit, ?int $limit = null): int
+    {
+        $subQueryBuilder = $this->entityManager->createQueryBuilder();
+        $subQueryBuilder->select('a.id')
+            ->from(AuditEventEntityInterface::class, 'a')
+            ->where('a.date < :latestDate')
+            ->setParameter('latestDate', $dateLimit->format(VUFIND_DATABASE_DATETIME_FORMAT));
+        if ($limit) {
+            $subQueryBuilder->setMaxResults($limit);
+        }
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder->delete(AuditEventEntityInterface::class, 'a')
+            ->where('a.id IN (:ids)')
+            ->setParameter('ids', $subQueryBuilder->getQuery()->getResult());
+        return $queryBuilder->getQuery()->execute();
+    }
+
+    /**
+     * Purge all events from the database.
+     *
+     * @return void
+     */
+    public function purgeEvents(): void
+    {
+        $this->entityManager->createQuery('DELETE ' . AuditEventEntityInterface::class . ' e')
+            ->execute();
+    }
+
+    /**
+     * Get the method that called the parent method here.
+     *
+     * @return string
+     */
+    protected function getCallerOfParentMethod(): string
+    {
+        $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 3);
+        $class = $backtrace[2]['class'] ?? '';
+        $function = $backtrace[2]['function'] ?? '';
+        return "$class::$function";
```

Comment:
Would this be a little neater in terms of avoiding a stray `::` if one or both parts are missing?

```suggestion
        $methodParts = [$backtrace[2]['class'] ?? '', $backtrace[2]['function'] ?? ''];
        return implode('::', array_filter($methodParts));
```